python/pyarrow/parquet/core.py

```diff
@@ -3423,13 +3427,15 @@ def write_metadata(schema, where, metadata_collector=None, **kwargs):
     ...     table.schema, 'dataset_metadata/_metadata',
     ...     metadata_collector=metadata_collector)
     """
-    writer = ParquetWriter(where, schema, **kwargs)
+    filesystem, where = _resolve_filesystem_and_path(where, filesystem)
+
+    writer = ParquetWriter(where, schema, filesystem, **kwargs)
     writer.close()
 
     if metadata_collector is not None:
         # ParquetWriter doesn't expose the metadata until it's written. Write
         # it and read it again.
-        metadata = read_metadata(where)
+        metadata = read_metadata(where, filesystem=filesystem)
         for m in metadata_collector:
             metadata.append_row_groups(m)
         metadata.write_metadata_file(where)
```

Comment:
I think we still need the `if filesystem is not None: with filesystem.open_output_stream(where) as f: ...` as you had before? 

(now, if that's correct, that also means the current tests are not covering this. Maybe it would be good to add a test using S3 filesystem, the current tests are still all local file paths, which will work without filesystem)

Reply:
Removing the `if filesystem is not None` at the top of this function then resolved the previous error for the file URI when `write_metadata_file` was not able to open file "file:///...".  But suppose here, you're right. For that point, I've added back that condition and updated the test. I can go about adding the s3 variant as well.

Reply:
Not sure if https://github.com/apache/arrow/pull/14574/commits/b8b9ceacd7adb020ee7c82426c4109c121f33769 is what you had it mind, let me know should anything be adjusted there. :)